cloud_detect/__init__.py

```diff
@@ -7,26 +10,65 @@
 from cloud_detect.providers import GCPProvider
 from cloud_detect.providers import OCIProvider
 
+__PROVIDER_CLASSES = [
+    AlibabaProvider, AWSProvider, AzureProvider,
+    DOProvider, GCPProvider, OCIProvider,
+]
 
-def provider(excluded=[]):
-    if 'alibaba' not in excluded and AlibabaProvider().identify():
-        logging.debug('Cloud_detect result is alibaba')
-        return 'alibaba'
-    elif 'aws' not in excluded and AWSProvider().identify():
-        logging.debug('Cloud_detect result is aws')
-        return 'aws'
-    elif 'gcp' not in excluded and GCPProvider().identify():
-        logging.debug('Cloud_detect result is gcp')
-        return 'gcp'
-    elif 'do' not in excluded and DOProvider().identify():
-        logging.debug('Cloud_detect result is do')
-        return 'do'
-    elif 'azure' not in excluded and AzureProvider().identify():
-        logging.debug('Cloud_detect result is azure')
-        return 'azure'
-    elif 'oci' not in excluded and OCIProvider().identify():
-        logging.debug('Cloud_detect result is oci')
-        return 'oci'
+TIMEOUT = 5  # seconds
+
+
+async def _identify(timeout):
+    tasks = {
+        prov.identifier: asyncio.ensure_future(
+            prov().identify(),
+        ) for prov in __PROVIDER_CLASSES
+    }
+
+    async def cancel_unfinished_tasks():
+        for t in tasks.values():
+            if not t.done():
+                try:
+                    t.cancel()
+                except asyncio.CancelledError:
+                    pass
+        # This statement ensures
+        # "Task was destroyed but it is pending!" warning is not raised
+        await asyncio.gather(*tasks.values())
+
+    stoptime = time.time() + timeout
+    while tasks and time.time() < stoptime:
+        for prov in list(tasks):
+            t = tasks[prov]
+            if t.done():
+                del tasks[prov]
+                if t.result():
+                    await cancel_unfinished_tasks()
+                    logging.debug('Cloud_detect result is %s' % prov)
+                    return prov
+        else:
+            await asyncio.sleep(0.1)
+            continue
+
+    if tasks:
+        await cancel_unfinished_tasks()
+        return 'timeout'
```

Comment:
If the code is run in a non-supported cloud environment or in local, the response is usually going to be`timeout` because almost all the cloud provider urls take more than 5s to respond. I am wondering if we should return `unknown in this case?

Reply:
In my latest commits I have set `timeout`'s default value as `None` and removed "timeout" as a response.
So, by default, `provider` will take all its time to determine the cloud provider. In my local, it took around 1 min 15s to return `unknown`.